analyzer/src/main/kotlin/managers/NuGet.kt

```diff
@@ -95,3 +104,59 @@ class NuGetPackageFileReader : XmlPackageFileReader {
         return ids
     }
 }
+
+/**
+ * A reader for XML-based NuGet configuration files, see
+ * https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file
+ */
+class NuGetConfigFileReader {
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @XmlRootElement(name = "configuration")
+    private data class NuGetConfig(
+        // Contains key:value pairs as well as the `<clear />` tag to ignore previously defined configuration values.
+        val packageSources: List<SortedMap<String, String>>
+    )
+
+    fun getRegistrationBaseUrls(configFile: File): List<String> {
+        val urls = mutableListOf<String>()
+        val nuGetConfig = NuGetSupport.XML_MAPPER.readValue<NuGetConfig>(configFile)
+        nuGetConfig.packageSources
+            .mapNotNull { it["value"] }
+            .filter {
+                if (it.startsWith("http")) {
+                    true
+                } else {
+                    // TODO: Handle local package sources
+                    log.warn { "Local NuGet package source $it was ignored." }
+                    false
+                }
+            }.forEach { urls += it }
+
+        return urls
+    }
+}
+
+private fun File.searchUpwardsForFileCaseInsensitive(searchFileName: String): File? {
+    if (!isDirectory) return null
+
+    var currentDir: File? = absoluteFile
+    while (currentDir != null && !fileExistsCaseInsensitive(currentDir, searchFileName)) {
+        currentDir = currentDir.parentFile
+    }
+
+    return currentDir?.let { dir ->
+        resolveFileCaseInsensitive(dir, searchFileName)
+    }
+}
+
+private fun fileExistsCaseInsensitive(dir: File, fileName: String): Boolean {
+    val files = dir.list() ?: return false
+
+    return files.any { it.equals(fileName, ignoreCase = true) }
```

Comment:
Note that `list()` also returns directories, but we only want files, so we should also check for `isFile` here. I'd do that check, which involves a conversion to `File`, only after the first cheaper check succeeds, i.e.

    return files.any { it.equals(fileName, ignoreCase = true) && dir.resolve(it).isFile }

Reply:
But then again, why not change the function to return a nullable `File` which points to the match, if any, so you can completely omit `resolveFileCaseInsensitive` below?

Reply:
Or actually the opposite: Keep *only* `resolveFileCaseInsensitive` and make use of it in `searchUpwardsForFileCaseInsensitive` already in the `while` loop?